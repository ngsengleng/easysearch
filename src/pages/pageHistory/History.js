import React, { useState, useEffect } from "react";

import RenderHistory from "../../components/RenderHistory";
import {
  Typography,
  makeStyles,
  Paper,
  Divider,
  Button,
  Box,
} from "@material-ui/core";
import { Pagination } from "@material-ui/lab";
import { db } from "../../config/firebase";
import { useAuth } from "../../context/AuthContext";

const useStyles = makeStyles((theme) => ({
  title: {
    margin: "auto",
    textAlign: "center",
  },
  tableBox: {
    [theme.breakpoints.down("xs")]: {
      width: "100%",
    },
    [theme.breakpoints.up("sm")]: {
      width: "50%",
    },
    margin: "auto",
    marginTop: "50px",
  },

  buttonBox: {
    margin: "30px",
    textAlign: "center",
    marginTop: "10px",
  },
  button: {
    backgroundColor: "#f44336",
    color: "#ffebee",
  },
}));
export default function History() {
  const classes = useStyles();
  const [searchHistory, setSearchHistory] = useState([]);
  const [currentPage, setCurrentPage] = useState(1);
  const itemsPerPage = 10;
  const [width, setWidth] = useState(window.innerWidth);
  const breakpoint = 750;
  const { currentUser } = useAuth();
  const [toBeDeleted, setToBeDeleted] = useState([]);
  const [isDeletingItems, setIsDeletingItems] = useState(false);

  useEffect(() => {
    // correct way to set title is with react-helmet (its a library, go look it up)
    document.title = "History";
    const handleWindowResize = () => setWidth(window.innerWidth);
    window.addEventListener("resize", handleWindowResize);

    const fetchSearchHistory = async () => {
      const allData = await db
        .collection("users")
        .doc(currentUser.uid)
        .collection("searchHistory")
        .get();
      const foodData = await db
        .collection("users")
        .doc(currentUser.uid)
        .collection("foodSearchHistory")
        .get();
      const shopArr = allData.docs.map((doc) => [doc.id, doc.data(), "shop"]);
      const foodArr = foodData.docs.map((doc) => [doc.id, doc.data(), "food"]);
      const arr = shopArr.concat(foodArr);
      arr.sort((a, b) => (a[1]["timestamp"] < b[1]["timestamp"] ? 1 : -1));
      setSearchHistory(arr);
    };

    fetchSearchHistory();
    return () => window.removeEventListener("resize", handleWindowResize);
  }, [currentUser]);

  const handleChange = (e, value) => {
    setCurrentPage(value);
  };

  const addToDelete = (item, type) => {
    setToBeDeleted([...toBeDeleted, [item, type]]);
  };

  const removeFromDelete = (item, type) => {
    const newDelete = toBeDeleted.filter((x) =>
      x[0] !== item ? true : x[1] !== type
    );
    setToBeDeleted(newDelete);
  };

  const resetDelete = () => {
    setIsDeletingItems(false);
    setToBeDeleted([]);
  };

  const indexOfLastItem = currentPage * itemsPerPage;
  const indexOfFirstItem = indexOfLastItem - itemsPerPage;
  const currentItems = searchHistory.slice(indexOfFirstItem, indexOfLastItem);
  return (
    <div>
      <Paper className={classes.tableBox}>
        {width < breakpoint ? (
          <Typography variant="h6" className={classes.title}>
            Showing {indexOfFirstItem + 1} -{" "}
            {indexOfLastItem > searchHistory.length
              ? searchHistory.length
              : indexOfLastItem}{" "}
            of {searchHistory.length} items
          </Typography>
        ) : (
          <Typography variant="h4" className={classes.title}>
            Showing {indexOfFirstItem + 1} -{" "}
            {indexOfLastItem > searchHistory.length
              ? searchHistory.length
              : indexOfLastItem}{" "}
            of {searchHistory.length} items
          </Typography>
        )}

        <Divider />
        <RenderHistory
          items={currentItems}
          isDeletingItems={isDeletingItems}
          addToDelete={addToDelete}
          removeFromDelete={removeFromDelete}
        />
        <Pagination
          onChange={handleChange}
          count={Math.ceil(searchHistory.length / itemsPerPage)}
          defaultPage={1}
        />
      </Paper>
      <Box className={classes.buttonBox}>
        {isDeletingItems ? (
          <>
            <Button
              color="primary"
              variant="outlined"
              onClick={() => resetDelete()}
            >
              stop deleting
            </Button>
            <Button color="primary" variant="outlined">
              delete all
            </Button>
          </>
        ) : (
          <Button
            color="inherit"
            variant="contained"
            onClick={() => setIsDeletingItems(!isDeletingItems)}
            className={classes.button}
          >
            delete search history
          </Button>
        )}
      </Box>
    </div>
  );
}
