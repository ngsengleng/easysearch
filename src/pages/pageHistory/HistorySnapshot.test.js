import React from "react";
import renderer from "react-test-renderer";
import History from "./History";
import { AuthProvider } from "../../context/AuthContext";
it("renders correctly", () => {
  const tree = renderer
    .create(
      <AuthProvider>
        <History />
      </AuthProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
