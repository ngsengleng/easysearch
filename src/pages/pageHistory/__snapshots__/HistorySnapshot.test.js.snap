// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    className="MuiPaper-root makeStyles-tableBox-2 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <h4
      className="MuiTypography-root makeStyles-title-1 MuiTypography-h4"
    >
      Showing 
      1
       -
       
      0
       
      of 
      0
       items
    </h4>
    <hr
      className="MuiDivider-root"
    />
    <ul
      className="MuiList-root makeStyles-list-5 MuiList-padding"
    />
    <nav
      aria-label="pagination navigation"
      className="MuiPagination-root"
    >
      <ul
        className="MuiPagination-ul"
      >
        <li>
          <button
            aria-label="Go to previous page"
            className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page Mui-disabled Mui-disabled"
            disabled={true}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={-1}
            type="button"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiPaginationItem-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
              />
            </svg>
          </button>
        </li>
        <li>
          <button
            aria-label="Go to next page"
            className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page Mui-disabled Mui-disabled"
            disabled={true}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={-1}
            type="button"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiPaginationItem-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
              />
            </svg>
          </button>
        </li>
      </ul>
    </nav>
  </div>
  <div
    className="MuiBox-root MuiBox-root-6 makeStyles-buttonBox-3"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-4 MuiButton-colorInherit"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        delete search history
      </span>
    </button>
  </div>
</div>
`;
